astropy/io/fits/src/compressionmodule.c

```diff
@@ -386,13 +386,25 @@ void tcolumns_from_header(fitsfile* fileptr, PyObject* header,
     long long totalwidth;
     int status = 0;
 
-    get_header_int(header, "TFIELDS", &tfields, 0);
+    if (get_header_int(header, "TFIELDS", &tfields, 0) == -1) {
+        return;
+    }
+    /* To avoid issues in the loop we need to limit the number of TFIELDs to
+       999. Otherwise we would exceed the maximum length of the keyword name of
+       8. This could lead to multiple accesses of the same header keyword with
+       snprintf because we limit it to 8 characters + null-termination. */
+    if (tfields > 999) {
+        PyErr_SetString(PyExc_ValueError, "The TFIELDS value exceeds 999.");
+        return;
+    }
 
     // This used to use PyMem_New, but don't do that; CFITSIO will later
     // free() this object when the file is closed, so just use malloc here
     // *columns = column = PyMem_New(tcolumn, (size_t) tfields);
     *columns = column = calloc((size_t) tfields, sizeof(tcolumn));
     if (column == NULL) {
+        PyErr_SetString(PyExc_MemoryError,
```

Comment:
If `calloc` fails we need to set the exception. Otherwise it would be a `SystemError` to return `NULL` without an Exception.

Reply:
hm, not sure if that makes much sense, no c or cfitsio function will set an exception so I would have to include that in a lot of places. I only mentioned it because `PyMem_New` would've set an exception. Probably just was forgotten when it was changed.